ligo/swaps.mligo

```diff
@@ -91,10 +92,11 @@ let rec y_to_x_rec (p : y_to_x_rec_param) : y_to_x_rec_param =
             (* The cached price corresponding to hi. *)
             let next_tick = get_tick p.s.ticks i_u in
             let sqrt_price_new = next_tick.sqrt_price in
-            (* How much dx will we receive for going all the wax to lo. *)
 
-            (* FIXME this is wrong, invert prices etc *)
-            let dx = Bitwise.shift_right (p.s.liquidity * (assert_nat (sqrt_price_new - p.s.sqrt_price))) 90n in
+            (* How much dx will we receive for going all the way to lo. *)
+            (* From 6.16 formula: dx = L * (1 / old sqrt_price - 1 / new sqrt_price), where dx is how X decreases *)
+            let dx = ceildiv (p.s.liquidity * Bitwise.shift_left (assert_nat (sqrt_price_new - p.s.sqrt_price)) 90n)
+                             (sqrt_price_new * p.s.sqrt_price) in
```

Comment:
I think we also have to shift `sqrt_price_new * p.s.sqrt_price` right by 180.

Reply:
If we decrease the divisible by `2^90` times and the divisor by `2^180` times, that would be the same as increasing the divisible by `2^90` times (modulo potential issues with precision loss) - does it make sense?

Or you have something different in mind? Could you write down the formula then?

Reply:
Oh yeah, good point! :smile: